File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/cafeteria/CafeteriaNotificationSettings.java
package de.tum.in.tumcampusapp.component.ui.cafeteria;

import android.content.Context;
import android.content.SharedPreferences;
import android.support.v4.util.Pair;
import android.support.v7.preference.PreferenceManager;

import org.joda.time.DateTime;
import org.joda.time.DateTimeConstants;
import org.joda.time.ReadableDateTime;

import java.util.ArrayList;
import java.util.Calendar;

import de.tum.in.tumcampusapp.component.ui.cafeteria.controller.CafeteriaMenuManager;

/**
 * This class takes care of saving a user's preferred time for notifying him about a cafeteria's
 * servings at a given day.
 */
public class CafeteriaNotificationSettings {
    private final SharedPreferences sharedPreferences;
    private final String PREFIX = "CAFETERIA_SCHEDULE_";
    private final String MINUTE = "_MINUTE";
    private final String HOUR = "_HOUR";
    private final CafeteriaMenuManager cafeteriaMenuManager;

    /**
     * Checks if there's already a preferred notification time for every weekday(Monday-Friday)
     * and for every day missing it defaults to DEFAULT_HOUR:DEFAULT_MINUTE
     */
    public CafeteriaNotificationSettings(Context context) {
        cafeteriaMenuManager = new CafeteriaMenuManager(context);
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        DateTime it = DateTime.now()
                .withDayOfWeek(DateTimeConstants.MONDAY);
        boolean wasAlreadyInitialized = true;
        while (it.getDayOfWeek() < DateTimeConstants.SATURDAY) {
            if (writeDayToSettings(it, 9, 30, false)) {
                wasAlreadyInitialized = false;
            }
            it = it.plusWeeks(1);
        }
        if (!wasAlreadyInitialized) {
            cafeteriaMenuManager.scheduleFoodAlarms(true);
        }
    }

    /**
     * If weekday is in range [Monday, Friday], set the preferred notification time for weekday
     * to hour:minute
     */
    private void writeDayToSettings(ReadableDateTime weekday, int hour, int minute) {
        int dayOfWeek = weekday.getDayOfWeek();
        if (dayOfWeek >= Calendar.MONDAY && dayOfWeek <= Calendar.FRIDAY) {
            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.putInt(PREFIX + dayOfWeek + HOUR, hour);
            editor.putInt(PREFIX + dayOfWeek + MINUTE, minute);
            editor.apply();
        }
    }

    /**
     * If weekday is in range [Monday, Friday], set the preferred notification time for weekday,
     * but only if, its either not set yet, or overwrite is set to true and the new hour:minute pair
     * differs from the one currently set
     */
    private boolean writeDayToSettings(ReadableDateTime weekday, int hour, int minute, boolean overwrite) {
        Pair<Integer, Integer> hourMinuteStored = retrieveHourMinute(weekday);
        if (hourMinuteStored == null) {
            writeDayToSettings(weekday, hour, minute);
        } else if (overwrite && (hourMinuteStored.first != hour || hourMinuteStored.second != minute)) {
            writeDayToSettings(weekday, hour, minute);
        } else {
            return false;
        }
        return true;
    }

    /**
     * Returns the currently stored preferred hour and minute of a given day of the week.
     *
     * @param day A day from the interval [MONDAY, FRIDAY]
     */

    public Pair<Integer, Integer> retrieveHourMinute(ReadableDateTime day) {
        int dayOfWeek = day.getDayOfWeek();
        if (dayOfWeek >= Calendar.MONDAY && dayOfWeek <= Calendar.FRIDAY) {
            int hour = sharedPreferences.getInt(PREFIX + dayOfWeek + HOUR, -2);
            int minute = sharedPreferences.getInt(PREFIX + dayOfWeek + MINUTE, -2);
            if (hour == -2 || minute == -2) {
                return null;
            }
            return new Pair<>(hour, minute);
        }
        return null;
    }

    /**
     * Checks if there's already a time preference for a weekday or not. The settingsPrefix
     * only differ by DAY_OF_WEEK.
     *
     * @param weekday A day ranging from MONDAY to FRIDAY
     * @param hour    The preferred hour for a notification
     * @param minute  The preferred minute for a notification
     */
    private boolean updateHourMinuteOfDay(DateTime weekday, int hour, int minute) {
        Pair<Integer, Integer> currentlyStored = retrieveHourMinute(weekday);
        if (currentlyStored != null && currentlyStored.first == hour && currentlyStored.second == minute) {
            return false;
        }
        return ((hour == -1 && minute == -1) || (hour >= 0 && hour < 24 && minute >= 0 && minute < 60)) && (writeDayToSettings(weekday, hour, minute, true));
    }

    /**
     * Stores all the preferred notification times from monday to friday.
     *
     * @param mondayToFriday An arraylist of hour-minute pairs ordered from monday to friday.
     * @return True if actual changes occured, False if nothing really changed.
     */
    public boolean saveWholeSchedule(ArrayList<Pair<Integer, Integer>> mondayToFriday) {
        if (mondayToFriday == null || mondayToFriday.size() != 5) {
            return false;
        } else {
            DateTime monday = DateTime.now();
            boolean storedSomething = false;
            for (int i = 0; i < mondayToFriday.size(); i++) {
                monday = monday.plusWeeks(1)
                        .withDayOfWeek(DateTimeConstants.MONDAY);
                if (updateHourMinuteOfDay(monday, mondayToFriday.get(i).first, mondayToFriday.get(i).second)) {
                    storedSomething = true;
                }
            }
            if (storedSomething) {
                cafeteriaMenuManager.scheduleFoodAlarms(true);
                return true;
            }
            return false;
        }
    }
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/cafeteria/CafeteriaNotificationSettings.kt
package de.tum.`in`.tumcampusapp.component.ui.cafeteria

import android.content.Context
import android.preference.PreferenceManager
import de.tum.`in`.tumcampusapp.component.ui.cafeteria.controller.CafeteriaMenuManager
import org.joda.time.DateTime
import org.joda.time.LocalTime

/**
 * This class encapsulates the user's settings for cafeteria notifications. It allows retrieval and
 * manipulation of notification times. It also reschedules alarms if notifications times are changed.
 *
 * @param context A [Context]
 */
class CafeteriaNotificationSettings(context: Context) {

    private val cafeteriaMenuManager = CafeteriaMenuManager(context)
    private val persistentStore = CafeteriaNotificationSettingsStore(context)

    /**
     * Returns the [LocalTime] of cafeteria notifications on a particular weekday, or null if the
     * user disabled notifications on this weekday.
     *
     * @param weekday The [DateTime] representing the weekday
     * @return The [LocalTime] of the notification or null
     */
    fun retrieveLocalTime(weekday: DateTime): LocalTime? {
        return persistentStore.retrieveTimeForDay(weekday)
    }

    /**
     * Returns the [LocalTime] of cafeteria notifications on a particular weekday. If the user
     * disabled notifications on this weekday, this method returns the default notification time.
     *
     * @param weekday The [DateTime] representing the weekday
     * @return The [LocalTime] of the notification or the default notification time
     */
    fun retrieveLocalTimeOrDefault(weekday: DateTime): LocalTime {
        return retrieveLocalTime(weekday) ?: defaultNotificationTime
    }

    /**
     * Stores the provided [CafeteriaNotificationTime].
     *
     * @param notificationTime The [CafeteriaNotificationTime] to store
     * @return Boolean whether the data in [CafeteriaNotificationSettingsStore] was updated
     */
    private fun updateLocalTimeOfDay(notificationTime: CafeteriaNotificationTime): Boolean {
        val currentlyStored = persistentStore.retrieveTimeForDay(notificationTime.weekday)

        if (notificationTime.time == currentlyStored) {
            return false
        }

        persistentStore.storeNotificationTime(notificationTime)
        return true
    }

    /**
     * Stores all [CafeteriaNotificationTime]s in the [CafeteriaNotificationSettingsStore] and
     * returns whether any time was actually changed.
     *
     * @param times The list of [CafeteriaNotificationTime]s
     * @return Boolean whether any time was actually changed
     */
    fun saveEntireSchedule(times: List<CafeteriaNotificationTime>): Boolean {
        if (times.size != 5) {
            return false
        }

        val didChangeTime = times
                .map { updateLocalTimeOfDay(it) }
                .toBooleanArray()
                .any { it }

        return if (didChangeTime) {
            cafeteriaMenuManager.scheduleNotificationAlarms()
            true
        } else {
            false
        }
    }

    companion object {

        private const val PREFIX = "CAFETERIA_SCHEDULE_"
        private const val HOUR_SUFFIX = "_HOUR"
        private const val MINUTE_SUFFIX = "_MINUTE"
        private const val NO_VALUE_SET = -1

        private var INSTANCE: CafeteriaNotificationSettings? = null

        private val defaultNotificationTime: LocalTime by lazy {
            LocalTime.now()
                    .withHourOfDay(9)
                    .withMinuteOfHour(30)
                    .withSecondOfMinute(0)
                    .withMillisOfSecond(0)
        }

        @JvmStatic
        @Synchronized
        fun getInstance(context: Context): CafeteriaNotificationSettings {
            if (INSTANCE == null) {
                INSTANCE = CafeteriaNotificationSettings(context)
            }

            return INSTANCE!!
        }
    }

    /**
     * This class is responsible for storing and retrieving [CafeteriaNotificationTime]s.
     *
     * @param context A [Context]
     */
    private class CafeteriaNotificationSettingsStore(context: Context) {

        private val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context)

        fun storeNotificationTime(notificationTime: CafeteriaNotificationTime) {
            storeTimeForDay(notificationTime.weekday.dayOfWeek, notificationTime.time)
        }

        private fun storeTimeForDay(dayOfWeek: Int, time: LocalTime?) {
            val hour = time?.hourOfDay ?: NO_VALUE_SET
            val minute = time?.minuteOfHour ?: NO_VALUE_SET
            sharedPrefs
                    .edit()
                    .putInt(PREFIX + dayOfWeek + HOUR_SUFFIX, hour)
                    .putInt(PREFIX + dayOfWeek + MINUTE_SUFFIX, minute)
                    .apply()
        }

        fun retrieveTimeForDay(day: DateTime) = retrieveTimeForDay(day.dayOfWeek)

        private fun retrieveTimeForDay(dayOfWeek: Int): LocalTime? {
            // The initial version of the app stored the hour and the minute value of the
            // notification time separately. To not break things, we continue to do so, but expose
            // the information as a LocalTime.
            val hour = sharedPrefs.getInt(PREFIX + dayOfWeek + HOUR_SUFFIX, NO_VALUE_SET)
            val minute = sharedPrefs.getInt(PREFIX + dayOfWeek + MINUTE_SUFFIX, NO_VALUE_SET)

            if (hour == NO_VALUE_SET || minute == NO_VALUE_SET) {
                return null
            }

            return LocalTime.now()
                    .withHourOfDay(hour)
                    .withMinuteOfHour(minute)
                    .withSecondOfMinute(0)
                    .withMillisOfSecond(0)
        }
    }
}
